From: Orit Wasserman <owasserm@redhat.com>
Date: Mon, 4 Jul 2016 15:01:51 +0200
Subject: rgw: remove bucket index objects when deleting the bucket

Fixes: http://tracker.ceph.com/issues/16412
Signed-off-by: Orit Wasserman <owasserm@redhat.com>
(cherry picked from commit 3ae276390641ad5fc4fef0c03971db95948880b4)

Resolves: rhbz#1353435
---
 src/rgw/rgw_rados.cc | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/src/rgw/rgw_rados.cc b/src/rgw/rgw_rados.cc
index 4992be0..24ffe36 100644
--- a/src/rgw/rgw_rados.cc
+++ b/src/rgw/rgw_rados.cc
@@ -7103,8 +7103,8 @@ bool RGWRados::is_syncing_bucket_meta(rgw_bucket& bucket)
 int RGWRados::delete_bucket(rgw_bucket& bucket, RGWObjVersionTracker& objv_tracker)
 {
   librados::IoCtx index_ctx;
-  string oid;
-  int r = open_bucket_index(bucket, index_ctx, oid);
+  map<int, string> bucket_objs;
+  int r = open_bucket_index(bucket, index_ctx, bucket_objs);
   if (r < 0)
     return r;
 
@@ -7144,6 +7144,11 @@ int RGWRados::delete_bucket(rgw_bucket& bucket, RGWObjVersionTracker& objv_track
     if (r < 0) {
       return r;
     }
+    /* remove bucket index objects*/
+    map<int, string>::const_iterator biter;
+    for (biter = bucket_objs.begin(); biter != bucket_objs.end(); ++biter) {
+      index_ctx.remove(biter->second);
+    }
   }
   return 0;
 }
